File: apps/client-macos/Sources/TimeTrack/Capture/ScreenshotScheduler.swift
import Foundation

/// PRD §6.2 — the capture trigger. A single self-gating interval timer (mirrors the heartbeat
/// timer's `guard isTracking`): each tick, only while the clock runs and only through `AckGate`,
/// it grabs the screen, writes to the durable image buffer, and kicks the upload drain. Capture is
/// tied to time tracking (design decision: no screenshots on a stopped clock). The interval and
/// enablement are an install-time snapshot — the caller builds this only when `screenshotsEnabled`,
/// and ack-revocation is caught per tick by the gate. Never captures on a closed gate; a missing
/// Screen Recording permission surfaces a menu-bar warning and the timer keeps running so capture
/// self-heals once granted.
///
/// The permission check (`isPermitted`) is a NON-prompting preflight (`CGPreflightScreenCaptureAccess`
/// via `ScreenRecordingPermission.isGranted`). It gates the grab so a denied tick never enters the
/// ScreenCaptureKit APIs — those re-trigger the one-time OS Screen Recording dialog on every call,
/// which made the prompt reappear each interval. Keep it a preflight; do not "simplify" it back into
/// the grab path.
final class ScreenshotScheduler {
    private let ackGate: AckGate
    private let grabber: DisplayGrabbing
    private let buffer: ImageBufferStore
    private let intervalSeconds: TimeInterval
    private let isTracking: () -> Bool
    private let isPermitted: () -> Bool
    private let idGen: (Date) -> String
    private let clock: () -> Date
    private let onCaptured: () -> Void
    private let onPermissionDenied: () -> Void
    private let onCaptureSucceeded: () -> Void

    private var timer: Timer?
    private var started = false
    private var isCapturing = false
    private var currentCycle: Task<Void, Never>?

    init(ackGate: AckGate, grabber: DisplayGrabbing, buffer: ImageBufferStore,
         intervalMinutes: Int, isTracking: @escaping () -> Bool,
         isPermitted: @escaping () -> Bool = ScreenRecordingPermission.isGranted,
         idGen: @escaping (Date) -> String = { UUIDv7.generate(now: $0) },
         clock: @escaping () -> Date = Date.init,
         onCaptured: @escaping () -> Void = {},
         onPermissionDenied: @escaping () -> Void = {},
         onCaptureSucceeded: @escaping () -> Void = {}) {
        self.ackGate = ackGate
        self.grabber = grabber
        self.buffer = buffer
        self.intervalSeconds = TimeInterval(intervalMinutes * 60)
        self.isTracking = isTracking
        self.isPermitted = isPermitted
        self.idGen = idGen
        self.clock = clock
        self.onCaptured = onCaptured
        self.onPermissionDenied = onPermissionDenied
        self.onCaptureSucceeded = onCaptureSucceeded
    }

    func start() {
        guard !started else { return }
        started = true
        scheduleNext()
    }

    func stop() {
        started = false
        timer?.invalidate()
        timer = nil
    }

    /// One capture attempt. `isTracking` is checked BEFORE the gate so a stopped clock never
    /// triggers a policy fetch. The whole grab runs inside `AckGate.withCaptureAllowed`.
    func captureTick() async {
        guard !isCapturing else { return }
        guard isTracking() else { return }
        isCapturing = true
        defer { isCapturing = false }
        do {
            try await ackGate.withCaptureAllowed { [self] in
                // Non-prompting preflight. If Screen Recording is denied, surface the warning and
                // skip the grab entirely — entering ScreenCaptureKit here would re-trigger the OS
                // permission dialog every interval. The timer keeps running, so capture self-heals
                // once the user grants access (onCaptureSucceeded clears the warning).
                guard isPermitted() else {
                    onPermissionDenied()
                    return
                }
                let capturedAt = clock()
                let id = idGen(capturedAt)
                let jpeg = try await grabber.grab()
                buffer.enqueue(id: id, capturedAt: capturedAt, jpeg: jpeg)
                onCaptureSucceeded()
                onCaptured()
            }
        } catch DisplayGrabError.notPermitted {
            // Fallback: permission was revoked between the preflight and the grab.
            onPermissionDenied()
        } catch {
            // Gate closed (ackRequired / offline) or another grab error → skip this tick.
        }
    }

    // MARK: - self-scheduling timer glue (build-verified)

    /// Spawn a capture cycle and hold it as `currentCycle` so sign-out teardown can join it via
    /// `finishInFlight()`. `internal` (not `private`) so the regression test can drive one cycle.
    func startCycle() {
        currentCycle = Task { [weak self] in
            await self?.captureTick()
            guard let self, self.started else { return }
            self.scheduleNext()
        }
    }

    /// Await any capture cycle already in flight (spawned before `stop()`). Sign-out teardown
    /// MUST await this before clearing the image buffer, or a capture suspended mid-grab could
    /// enqueue into the just-cleared buffer and upload under the next user's token.
    func finishInFlight() async { await currentCycle?.value }

    private func scheduleNext() {
        timer?.invalidate()
        let t = Timer(timeInterval: max(0.001, intervalSeconds), repeats: false) { [weak self] _ in
            self?.startCycle()
        }
        RunLoop.main.add(t, forMode: .common)
        timer = t
    }
}

File: apps/client-windows/src/NiftyTimer/Capture/ScreenshotScheduler.cs
using NiftyTimer.Policy;
using NiftyTimer.Storage;
using NiftyTimer.Tracking;

namespace NiftyTimer.Capture;

/// <summary>
/// PRD §6.2 — the screenshot capture trigger.
///
/// A single self-gating interval timer. Each tick, only while the clock is running and only
/// through <see cref="AckGate"/>, it grabs EVERY attached display, writes each frame to the
/// durable image buffer under one shared capture-group id, and lets the sync engine drain them.
///
/// Capture is tied to time tracking: there are no screenshots on a stopped clock. <c>isTracking</c>
/// is checked BEFORE the gate so a stopped clock never triggers a policy fetch.
///
/// The interval and whether screenshots run at all are an install-time snapshot — the caller
/// constructs this only when <c>screenshotsEnabled</c> — but acknowledgement is re-checked by the
/// gate on EVERY tick, so revoking it stops capture mid-session rather than at the next launch.
///
/// **No permission preflight, unlike the macOS counterpart.** That client checks
/// <c>CGPreflightScreenCaptureAccess</c> before every grab because entering ScreenCaptureKit
/// without the TCC grant re-triggers the OS permission dialog on each interval. Windows has no
/// equivalent gate on screen capture, so there is no permission to check, nothing to warn about,
/// and no self-healing path to build. Porting it as a <c>() =&gt; true</c> stub would leave dead
/// plumbing and a warning state that can never be entered.
/// </summary>
public sealed class ScreenshotScheduler : IDisposable
{
    private readonly AckGate _ackGate;
    private readonly IDisplayGrabber _grabber;
    private readonly IImageBuffer _buffer;
    private readonly TimeSpan _interval;
    private readonly Func<bool> _isTracking;
    private readonly Func<DateTimeOffset, string> _idGen;
    private readonly Func<DateTimeOffset, string> _groupIdGen;
    private readonly Func<DateTimeOffset> _clock;
    private readonly Action _onCaptured;
    private readonly Lock _gate = new();

    private CancellationTokenSource _cycles = new();
    private Timer? _timer;
    private Task _currentCycle = Task.CompletedTask;
    private bool _started;
    private bool _isCapturing;
    private bool _disposed;

    public ScreenshotScheduler(
        AckGate ackGate,
        IDisplayGrabber grabber,
        IImageBuffer buffer,
        int intervalMinutes,
        Func<bool> isTracking,
        Func<DateTimeOffset, string>? idGen = null,
        Func<DateTimeOffset, string>? groupIdGen = null,
        Func<DateTimeOffset>? clock = null,
        Action? onCaptured = null)
    {
        _ackGate = ackGate;
        _grabber = grabber;
        _buffer = buffer;
        _interval = TimeSpan.FromMinutes(Math.Max(1, intervalMinutes));
        _isTracking = isTracking;
        _idGen = idGen ?? (static now => UuidV7.Generate(now));
        _groupIdGen = groupIdGen ?? (static now => UuidV7.Generate(now));
        _clock = clock ?? (() => DateTimeOffset.UtcNow);
        _onCaptured = onCaptured ?? (static () => { });
    }

    public void Start()
    {
        lock (_gate)
        {
            if (_started || _disposed)
            {
                return;
            }

            _started = true;
        }

        ScheduleNext(_interval);
    }

    public void Stop()
    {
        CancellationTokenSource cancelled;
        lock (_gate)
        {
            _started = false;
            _timer?.Dispose();
            _timer = null;
            cancelled = _cycles;
            _cycles = new CancellationTokenSource();
        }

        cancelled.Cancel();
        cancelled.Dispose();
    }

    /// <summary>
    /// Await any capture cycle already in flight. Sign-out teardown MUST await this before
    /// clearing the image buffer: a grab suspended mid-cycle would otherwise resume afterwards and
    /// enqueue into the just-cleared buffer, uploading one user's screen under the next user's
    /// token.
    /// </summary>
    public Task FinishInFlightAsync()
    {
        lock (_gate)
        {
            return _currentCycle;
        }
    }

    /// <summary>
    /// One capture attempt. Returns true when frames were captured and buffered.
    /// </summary>
    public async Task<bool> CaptureTickAsync(CancellationToken cancellationToken = default)
    {
        lock (_gate)
        {
            if (_isCapturing)
            {
                return false;
            }

            _isCapturing = true;
        }

        try
        {
            if (!_isTracking())
            {
                return false;
            }

            return await _ackGate.WithCaptureAllowedAsync(GrabAsync, cancellationToken).ConfigureAwait(false);
        }
        catch (Exception e) when (e is AckGateException or Auth.NotAuthenticatedException
                                      or Auth.AuthException or DisplayGrabException
                                      or OperationCanceledException)
        {
            // Gate closed, session unusable, nothing grabbable, or torn down mid-tick. Skip the
            // tick; the timer keeps running so capture resumes on its own.
            return false;
        }
        finally
        {
            lock (_gate)
            {
                _isCapturing = false;
            }
        }
    }

    public void Dispose()
    {
        if (_disposed)
        {
            return;
        }

        _disposed = true;
        Stop();
        _cycles.Dispose();
    }

    private async Task<bool> GrabAsync(CancellationToken cancellationToken)
    {
        var capturedAt = _clock();

        // The fan-out across displays happens INSIDE the gate, and every frame shares one capture
        // time and one group id — so the dashboard reads a two-monitor desk as a single moment
        // rather than as unrelated screenshots taken seconds apart.
        var result = await _grabber.GrabAllAsync(cancellationToken).ConfigureAwait(false);
        if (result.Captures.Count == 0)
        {
            return false;
        }

        var groupId = _groupIdGen(capturedAt);
        foreach (var capture in result.Captures)
        {
            _buffer.Enqueue(
                _idGen(capturedAt),
                capturedAt,
                capture.Jpeg,
                new CaptureGroup(groupId, capture.Index, result.Attempted));
        }

        _onCaptured();
        return true;
    }

    private async Task RunCycleAsync(CancellationToken cancellationToken)
    {
        await CaptureTickAsync(cancellationToken).ConfigureAwait(false);

        lock (_gate)
        {
            if (!_started)
            {
                return;
            }
        }

        // Always the full interval, measured or skipped. Unlike the activity sampler, a tick here
        // costs a fraction of a second rather than consuming the interval, so there is nothing to
        // make contiguous.
        ScheduleNext(_interval);
    }

    private void ScheduleNext(TimeSpan delay)
    {
        lock (_gate)
        {
            if (!_started)
            {
                return;
            }

            var token = _cycles.Token;
            _timer?.Dispose();
            _timer = new Timer(
                _ =>
                {
                    lock (_gate)
                    {
                        _currentCycle = RunCycleAsync(token);
                    }
                },
                null,
                delay < TimeSpan.Zero ? TimeSpan.Zero : delay,
                Timeout.InfiniteTimeSpan);
        }
    }
}
